### ISSUE TITLE
What can I do to speed up EF Core Sqlite initialisation?

### ISSUE THREAD

**Issue by Metritutus** [aspnet/EntityFrameworkCore#12087]
In developing a Xamarin Android app using Entity Framework Core 2.0 with a code-first approach using a Sqlite database provider, the initialisation time seems pretty slow.  For approximately 40 models the app takes around 14 seconds to launch, 10 seconds for the call to `Migrate()`, and another 3-5 seconds for the first query after that.  Once those are done it seems to be fine.

I'm not sure if it relates to [#4372](https://github.com/aspnet/EntityFrameworkCore/issues/4372) or [#1906](https://github.com/aspnet/EntityFrameworkCore/issues/1906), but it may well do.

Is there anything I can do to help mitigate the initialisation time?

**Comment by bricelam** [comment]
Does keeping the connection open help? (It will a lot if you're using encryption.)
``` csharp
_connection = new SqliteConnection(connectionString);
_connection.Open();
optionsBuilder.UseSqlite(_connection);
```

**Comment by Metritutus** [comment]
Unfortunately I'm not using encryption and the change didn't seem to make much of a difference.

**Comment by cwrea** [comment]
@Metritutus In the mitigation category, try enabling SQLite's [Write-Ahead Logging (WAL)](https://www.sqlite.org/wal.html). I've found that it does reduce the time for `EnsureCreated()` in particular, but YMMV as I wasn't testing with a large number of model classes. See [here](https://github.com/cwrea/XamarinTodo/issues/1) for code example. My past experience with WAL in non-Xamarin mobile apps that use SQLite is that WAL does improve performance in a number of scenarios.

**Comment by Metritutus** [comment]
Sadly, enabling Write-Ahead-Logging did not appear to have any noticeable impact on initialisation time.

**Comment by tipa** [comment]
Any updates to this?
Many of my users are asking about why my app is starting so slowly compared to other apps in the store. Even with AOT enabled startup times are ~4 secs on high end phones.
This issue (together with [this one](https://github.com/aspnet/EntityFrameworkCore/issues/10963)) make me seriously consider making me switch to a more lightweight library

**Comment by ajcvickers** [comment]
@tipa Unfortunately we do not have the resources to spend time significant time on the Xamarin experience. While I hope that we get back to it in the future there are so many other issues that are higher priority that I think if you want something in the short term that is performant you would be better off using something other than EF.

**Comment by NickA55** [comment]
> @tipa Unfortunately we do not have the resources to spend time significant time on the Xamarin experience. While I hope that we get back to it in the future there are so many other issues that are higher priority that I think if you want something in the short term that is performant you would be better off using something other than EF.

This is really sad and disappointing to hear.  The future Microsoft is pushing is .NET Core, EF Core being part of that.  It begs the question of how does Xamarin fit into the plans?  Is Xamarin that low on the priority list that you really have no one to work on issues? 

This is a major issue, and makes EF Core on Xamarin totally unusable in a real world application.  So if I want to create a Xamarin application, and use what Microsoft wants me to use, which would be EF Core, I can't because it's so slow it's basically useless.  

This tells me a lot about what Microsoft really thinks of Xamarin as a serious development tool, and why Xamarin will never be taken seriously as a professional application.

**Comment by lnaie** [comment]
Come on guys, make every mobile developer proud of using Xamarin. :)

**Comment by dfoulk** [comment]
> @tipa Unfortunately we do not have the resources to spend time significant time on the Xamarin experience. While I hope that we get back to it in the future there are so many other issues that are higher priority that I think if you want something in the short term that is performant you would be better off using something other than EF.

Please excuse me for saying so, but this is a real bummer to hear. If this is a common theme across other teams- then I suppose it explains a lot of issues we've been having lately (getting Microsoft to support Xamarin).

@ajcvickers is this the way it's going to be or do you see any hope that Xamarin will be considered equal to other workloads in the future?

**Comment by Shumigaj** [comment]
My application also has slow initialization time because of EF Core. I've collected everything I learned in the [stackoverflow question](https://stackoverflow.com/questions/62582954/ef-core-sqlite-slow-star-up-performance-on-xamarin). I hope Microsoft team comes up with something before the release of EF Core 5.

**Comment by roji** [comment]
Am going to do my best to work on this for 5.0, but some sort of minimal code sample would greatly increase the chance of improving the perf here.

**Comment by Shumigaj** [comment]
Thanks, I'll provide clean example before Monday

**Comment by adelarsq** [comment]
The migration was the main issue for us (for almost 500 tables) even with no migration to be applied. So we had to remove EF from the app and use native. We use EF just to generate the sql files to execute on the app.

**Comment by Shumigaj** [comment]
Example with EF Core 5 Preview 6 added to [GitHub](https://github.com/Shumigaj/Samples/tree/1507d4682808b13072b6296d1fca0c3d0f0fd5a4). I also updated [stackoverflow question](https://stackoverflow.com/questions/62582954/ef-core-sqlite-slow-star-up-performance-on-xamarin) with additional information.

**Comment by ajcvickers** [comment]
Note from triage: we're still planning on doing this in the near future, but we are out of time to get significant changes into 5.0.

**Comment by dfyx** [comment]
Some of our users report up to 25 seconds when opening the first database context, especially on Android and older iOS devices.

**Comment by adelarsq** [comment]
Also more migrations added will turn to worst the performance to build and open the app.

**Comment by dfoulk** [comment]
Please don't punt this again. We need this fixed as EF Core migrations are invaluable in our web projects- but not currently viable in Xamarin. It's a tragedy, really... Thank you

**Comment by AndriySvyryd** [comment]
I haven't tested this, but if your project has a lot of migrations you could move them to a separate assembly and only call Migrate on install or update.

Model initialization perf will be addressed by https://github.com/dotnet/efcore/issues/1906 and https://github.com/dotnet/efcore/issues/19610

**Comment by adelarsq** [comment]
> I haven't tested this, but if your project has a lot of migrations you could move them to a separate assembly and only call Migrate/EnsureCreated on install or update.

Yes, this can help. I did exactly this. Also to avoid performance issues I search for migrations that exists on the migrations assembly that aren't on migration table. If there is any I call `Migrate()`.

**Comment by TomBruyneel** [comment]
Unfortunately the performance issues have little to do with migrations and everything with initializing the context the first time. So if your app needs some data from the database right on startup there is no way to escape from a slow startup. Reducing the number of entities also doesn't really help a lot.

See for example: https://stackoverflow.com/questions/62582954/ef-core-sqlite-slow-star-up-performance-on-xamarin

**Comment by dfyx** [comment]
I can confirm this. Taking measures to prevent the migrations from running when they aren't needed speeds up the process by about 20% but it's still a lot slower than it could be. On some Android devices and older iOS devices (especially those with A9 processors like the iPod Touch 5) it can take about 20 seconds to open an SQLCipher database while opening the database directly without EF it takes less than a second.

**Comment by dfyx** [comment]
As much as I hate pushing an old issue, this is still very much a problem for our customers (especially with SQLCipher where entering the wrong password means you need to sit through those 20 seconds multiple times). This issue was first reported two major versions and almost three years ago, it was punted multiple times, 5.0.0 was released months ago and there's still... nothing?

**Comment by TomBruyneel** [comment]
looking forward to 6.0 in that case

**Comment by dfoulk** [comment]
Will we see this issue in MAUI apps as well (if you don't fix it in 6.0)?

**Comment by ChaseKnowlden** [comment]
Xamarin is dead, so can we close this issue?

**Comment by roji** [comment]
I think this issue can now track MAUI performance instead.
